### PR TITLE
ogre_helpers: fix bad ascii stl import segfault

### PR DESCRIPTION
**Pull request by jbohren**

### CODE REVIEW

File: src/rviz/ogre_helpers/stl_loader.cpp [hunk 1]
```diff
@@ -91,6 +91,13 @@ bool STLLoader::load(const std::string& path)
 bool STLLoader::load(uint8_t* buffer)
 {
   uint8_t* pos = buffer;
+
+  // check for ascii type
+  if(std::string((char*)buffer,5) == std::string("solid")) {
```
Comment (wjwwood) [inline src/rviz/ogre_helpers/stl_loader.cpp]: nitpick: `, 5`

File: src/rviz/ogre_helpers/stl_loader.cpp [hunk 2]
```diff
@@ -91,6 +91,13 @@ bool STLLoader::load(const std::string& path)
...
+    ROS_ERROR("Trying to load STL file from binary buffer, but the given data is an ASCII STL file. Please convert it to a binary STL file instead.");
```
Comment (wjwwood) [inline src/rviz/ogre_helpers/stl_loader.cpp]: Another nitpick: This line is really long, can you please do something like this:

```c++
ROS_ERROR("Trying to load STL file from binary buffer, but the given data is an ASCII STL file. "
          "Please convert it to a binary STL file instead.");
```

**Comment by wjwwood** [comment]
When you open an issue, please provide some context about what the pull request does, why it is needed, and why the fix exists. I know it takes more time, but the subject plus a lack of a description would make it very hard for others to find this issue by searching. You could include the backtrace from `gdb`, or link to the Wikipedia page about STL (http://en.wikipedia.org/wiki/STL_%28file_format%29#ASCII_STL). If you don't have time to make a full description when you post it, just leave a note saying you're going to come back and flesh it out later.

Other than the comments, it lgtm.

File: src/rviz/ogre_helpers/stl_loader.cpp (hunk 1 above)
Comment (wjwwood) [inline src/rviz/ogre_helpers/stl_loader.cpp]: You don't have to change this, but you should consider avoiding using C-style casting in C++, it's almost always more appropriate to use one of the fixed casting types in C++, in this case:

```c++
if (std::string(reinterpret_cast<char *>(buffer), 5) == std::string("solid")) {
```

**Comment by jbohren** [comment]
Here's the segfault:

```
Program received signal SIGSEGV, Segmentation fault.
0x00007ffff7a69406 in ogre_tools::STLLoader::load (this=0x7fffffffa5b0, 
    buffer=0x7fff5bb9a010 "solid ascii\n facet normal -0.616194 0.734641 0.283914\n  outer loop\n   vertex -0.027198 0.026293 0.025324\n   vertex -0.026537 0.026694 0.025721\n   vertex -0.026726 0.026983 0.024563\n  endloop\n endfacet"...)
    at /home/jbohren/ws/ascent/src/rviz/src/rviz/ogre_helpers/stl_loader.cpp:110
110	    tri.vertices_[0].x = *(float*)pos;
```

File: src/rviz/ogre_helpers/stl_loader.cpp (hunk 1 above)
Comment (wjwwood) [inline src/rviz/ogre_helpers/stl_loader.cpp]: Actually I think that can (should) be a `static_cast`.

Comment (caguero) [inline src/rviz/ogre_helpers/stl_loader.cpp]: This should be `if(std::string((char*)buffer,5) != std::string("solid")) {`

**Comment by wjwwood** [comment]
Closing in favor of: https://github.com/ros-visualization/rviz/pull/871

File: src/rviz/ogre_helpers/stl_loader.cpp (hunk 1 above)
Comment (jbohren) [inline src/rviz/ogre_helpers/stl_loader.cpp]: @caguero Maybe wording the error message is confusing, but there is an error if the STL file is an ASCII file. So we want to report an error if the file is in ASCII format, i.e. if the first 5 chars read as "solid".

**Comment by jbohren** [comment]
> Closing in favor of: #871 

Ok, well at least #871 uses the preferred casting syntax, even if it breaks rviz...

**Comment by wjwwood** [comment]
@jbohren Well, it's hard to know what you meant the patch to do with out a detailed description of the issue. We took it to mean that you were expecting a an ASCII STL file, and that if it did not start with `solid` that it was not a valid STL. We we tested it with a file that was not an STL, and the logic was reversed based on that check. I'll revert the patch.

**Comment by davetcoleman** [comment]
@jbohren thanks for putting in the effort to share this fix back into rviz. many other people wouldn't even bother or they would just open a ROS answers cry for help

**Pull request closed without merging**